Build failure: rpcs3

### Nixpkgs version

- 1ec9dbf3abfcd01a520a5a19a25d60c2ab28fac4 (master as of right now)

### Steps to reproduce

1. `nix build github:NixOS/nixpkgs/1ec9dbf3abfcd01a520a5a19a25d60c2ab28fac4#rpcs3 -L`

### Can Hydra reproduce this build failure?

Search doesn't work for me so idk.

### Relevant log output

```console
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp: In lambda function:
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp:2339:61: error: use of deleted function 'atomic_t<T, Align>::atomic_t(const atomic_t<T, Align>&) [with T = int; long unsigned int Align = 4]'
rpcs3>  2339 |                         pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
rpcs3>       |                                            ~~~~~~~~~~~~~~~~~^~~~~~~~
rpcs3> In file included from /build/source/rpcs3/rpcs3qt/movie_item_base.h:4,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/game_list_base.h:3,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/game_list.h:9,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/game_list_frame.h:3,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/game_list_frame.cpp:1:
rpcs3> /build/source/rpcs3/util/atomic.hpp:1192:9: note: declared here
rpcs3>  1192 |         atomic_t(const atomic_t&) = delete;
rpcs3>       |         ^~~~~~~~
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp:2339:61: note: use '8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/C_002b_002b-Dialect-Options.html#index-fdiagnostics-all-candidates-fdiagnostics-all-candidates8;;' to display considered candidates
rpcs3>  2339 |                         pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
rpcs3>       |                                            ~~~~~~~~~~~~~~~~~^~~~~~~~
rpcs3> In file included from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtGui/qkeysequence.h:8,
rpcs3>                  from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtGui/qaction.h:9,
rpcs3>                  from /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtGui/QAction:1,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/game_list.h:3:
rpcs3> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:316:33: note:   initializing argument 1 of 'QString QString::arg(T, int, int, QChar) const [with T = atomic_t<int, 4>; typename std::enable_if<conjunction_v<std::disjunction<std::is_convertible<T, long long unsigned int>, std::is_convertible<T, long long int> >, std::negation<std::disjunction<std::is_same<typename std::remove_cvref<_Tp>::type, _Float16>, std::is_same<typename std::remove_cvref<_Tp>::type, qfloat16>, std::is_floating_point<_Tp> > >, std::negation<std::conjunction<std::negation<QtPrivate::treat_as_integral_arg<typename std::remove_cv< <template-parameter-1-1> >::type> >, std::is_convertible<T, QAnyStringView> > > >, bool>::type <anonymous> = true]'
rpcs3>   316 |     [[nodiscard]] QString arg(T a, int fieldWidth = 0, int base = 10,
rpcs3>       |                               ~~^
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp: In lambda function:
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp:2372:53: error: use of deleted function 'atomic_t<T, Align>::atomic_t(const atomic_t<T, Align>&) [with T = int; long unsigned int Align = 4]'
rpcs3>  2372 |                 pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
rpcs3>       |                                    ~~~~~~~~~~~~~~~~~^~~~~~~~
rpcs3> /build/source/rpcs3/util/atomic.hpp:1192:9: note: declared here
rpcs3>  1192 |         atomic_t(const atomic_t&) = delete;
rpcs3>       |         ^~~~~~~~
rpcs3> /build/source/rpcs3/rpcs3qt/game_list_frame.cpp:2372:53: note: use '8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/C_002b_002b-Dialect-Options.html#index-fdiagnostics-all-candidates-fdiagnostics-all-candidates8;;' to display considered candidates
rpcs3>  2372 |                 pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
rpcs3>       |                                    ~~~~~~~~~~~~~~~~~^~~~~~~~
rpcs3> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:316:33: note:   initializing argument 1 of 'QString QString::arg(T, int, int, QChar) const [with T = atomic_t<int, 4>; typename std::enable_if<conjunction_v<std::disjunction<std::is_convertible<T, long long unsigned int>, std::is_convertible<T, long long int> >, std::negation<std::disjunction<std::is_same<typename std::remove_cvref<_Tp>::type, _Float16>, std::is_same<typename std::remove_cvref<_Tp>::type, qfloat16>, std::is_floating_point<_Tp> > >, std::negation<std::conjunction<std::negation<QtPrivate::treat_as_integral_arg<typename std::remove_cv< <template-parameter-1-1> >::type> >, std::is_convertible<T, QAnyStringView> > > >, bool>::type <anonymous> = true]'
rpcs3>   316 |     [[nodiscard]] QString arg(T a, int fieldWidth = 0, int base = 10,
rpcs3>       |                               ~~^
rpcs3> [ 87%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/localized_emu.cpp.o
rpcs3> [ 87%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/log_frame.cpp.o
rpcs3> [ 87%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/log_viewer.cpp.o
rpcs3> [ 87%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/main_window.cpp.o
rpcs3> [ 88%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/memory_string_searcher.cpp.o
rpcs3> [ 88%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/memory_viewer_panel.cpp.o
rpcs3> [ 88%] Building CXX object rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/microphone_creator.cpp.o
rpcs3> make[2]: *** [rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/build.make:558: rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/game_list_frame.cpp.o] Error 1
rpcs3> make[2]: *** Waiting for unfinished jobs....
rpcs3> In file included from /nix/store/pdcaxxfvs5smr8pv1xkypmbxqrq6qqdh-wolfssl-all-5.7.4-dev/include/wolfssl/ssl.h:33,
rpcs3>                  from /build/source/rpcs3/Emu/NP/rpcn_client.h:37,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/rpcn_settings_dialog.h:9,
rpcs3>                  from /build/source/rpcs3/rpcs3qt/main_window.cpp:13:
rpcs3> /nix/store/pdcaxxfvs5smr8pv1xkypmbxqrq6qqdh-wolfssl-all-5.7.4-dev/include/wolfssl/wolfcrypt/settings.h:3601:14: warning: #warning "For timing resistance / side-channel attack prevention consider using harden options" [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wcpp-Wcpp8;;]
rpcs3>  3601 |             #warning "For timing resistance / side-channel attack prevention consider using harden options"
rpcs3>       |              ^~~~~~~
rpcs3> make[1]: *** [CMakeFiles/Makefile2:2070: rpcs3/rpcs3qt/CMakeFiles/rpcs3_ui.dir/all] Error 2
rpcs3> make: *** [Makefile:156: all] Error 2
```

### System metadata

- system: `"x86_64-linux"`
 - host os: `Linux 6.12.23-xanmod1, NixOS, 25.05 (Warbler), 25.05.20250421.c11863f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/nat/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/nat/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/nat/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/nat/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf:/nix/store/pz2l9m282mx9aqyskd1pjs3rh4sgyk93-gnome-settings-daemon-47.2/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/jxs248g15qklggm9gfyddq822vkcrfg2-lix-2.91.1/share`
 - nixpkgs: `/nix/store/kcmmd6alr3lx56vkf72503h3pxgf6iv4-source`

### Notify maintainers

@abbradar @neonfuz @ilian @zane

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/rp/rpcs3/fix-qt6.9-compilation.patch
+++ b/pkgs/by-name/rp/rpcs3/fix-qt6.9-compilation.patch
@@ -1,0 +1,68 @@
+From 600e4604169464c64cbf548e7629e483ad2aad1e Mon Sep 17 00:00:00 2001
+From: Megamouse <studienricky89@googlemail.com>
+Date: Mon, 7 Apr 2025 20:49:15 +0200
+Subject: [PATCH] Fix compilation with newer Qt
+
+---
+ rpcs3/rpcs3qt/game_list_base.cpp         | 2 +-
+ rpcs3/rpcs3qt/game_list_frame.cpp        | 4 ++--
+ rpcs3/rpcs3qt/ps_move_tracker_dialog.cpp | 4 ++--
+ 3 files changed, 5 insertions(+), 5 deletions(-)
+
+diff --git a/rpcs3/rpcs3qt/game_list_base.cpp b/rpcs3/rpcs3qt/game_list_base.cpp
+index 72b45d33bfbb..21640906ee7e 100644
+--- a/rpcs3/rpcs3qt/game_list_base.cpp
++++ b/rpcs3/rpcs3qt/game_list_base.cpp
+@@ -25,7 +25,7 @@ void game_list_base::repaint_icons(std::vector<game_info>& game_data, const QCol
+ 	for (game_info& game : game_data)
+ 	{
+ 		game->pxmap = placeholder;
+-	
++
+ 		if (movie_item_base* item = game->item)
+ 		{
+ 			item->set_icon_load_func([this, game, device_pixel_ratio, cancel = item->icon_loading_aborted()](int)
+diff --git a/rpcs3/rpcs3qt/game_list_frame.cpp b/rpcs3/rpcs3qt/game_list_frame.cpp
+index a294d69cc68e..439913760053 100644
+--- a/rpcs3/rpcs3qt/game_list_frame.cpp
++++ b/rpcs3/rpcs3qt/game_list_frame.cpp
+@@ -2363,7 +2363,7 @@ void game_list_frame::BatchActionBySerials(progress_dialog* pdlg, const std::set
+ 
+ 		connect(future_watcher, &QFutureWatcher<void>::finished, this, [=, this]()
+ 		{
+-			pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
++			pdlg->setLabelText(progressLabel.arg(index->load()).arg(serials_size));
+ 			pdlg->setCancelButtonText(tr("OK"));
+ 			QApplication::beep();
+ 
+@@ -2396,7 +2396,7 @@ void game_list_frame::BatchActionBySerials(progress_dialog* pdlg, const std::set
+ 			return;
+ 		}
+ 
+-		pdlg->setLabelText(progressLabel.arg(*index).arg(serials_size));
++		pdlg->setLabelText(progressLabel.arg(index->load()).arg(serials_size));
+ 		pdlg->setCancelButtonText(tr("OK"));
+ 		connect(pdlg, &progress_dialog::canceled, this, [pdlg](){ pdlg->deleteLater(); });
+ 		QApplication::beep();
+diff --git a/rpcs3/rpcs3qt/ps_move_tracker_dialog.cpp b/rpcs3/rpcs3qt/ps_move_tracker_dialog.cpp
+index 45fbe6f59e7d..4b9bc5dd6e4f 100644
+--- a/rpcs3/rpcs3qt/ps_move_tracker_dialog.cpp
++++ b/rpcs3/rpcs3qt/ps_move_tracker_dialog.cpp
+@@ -362,7 +362,7 @@ void ps_move_tracker_dialog::update_saturation_threshold(bool update_slider)
+ }
+ void ps_move_tracker_dialog::update_min_radius(bool update_slider)
+ {
+-	ui->minRadiusGb->setTitle(tr("Min Radius: %0 %").arg(g_cfg_move.min_radius));
++	ui->minRadiusGb->setTitle(tr("Min Radius: %0 %").arg(g_cfg_move.min_radius.get()));
+ 
+ 	if (update_slider)
+ 	{
+@@ -372,7 +372,7 @@ void ps_move_tracker_dialog::update_min_radius(bool update_slider)
+ 
+ void ps_move_tracker_dialog::update_max_radius(bool update_slider)
+ {
+-	ui->maxRadiusGb->setTitle(tr("Max Radius: %0 %").arg(g_cfg_move.max_radius));
++	ui->maxRadiusGb->setTitle(tr("Max Radius: %0 %").arg(g_cfg_move.max_radius.get()));
+ 
+ 	if (update_slider)
+ 	{

--- a/pkgs/by-name/rp/rpcs3/package.nix
+++ b/pkgs/by-name/rp/rpcs3/package.nix
@@ -58,6 +58,11 @@ stdenv.mkDerivation {
     hash = rpcs3Hash;
   };
 
+  patches = [
+    # Modified from https://github.com/RPCS3/rpcs3/pull/17009; doesn't apply cleanly due to intermediate commits
+    ./fix-qt6.9-compilation.patch
+  ];
+
   passthru.updateScript = ./update.sh;
 
   preConfigure = ''

